keyboards/preonic/keymaps/snowskeleton/keymap.c

```diff
@@ -0,0 +1,237 @@
+ * `-----------------------------------------------------------------------------------'
+ */
+[_ADJUST] = LAYOUT_preonic_grid(
+  _______, _______, _______,   _______, _______, _______, _______, _______, _______,   _______, _______, _______,
+  _______, QK_BOOT, DB_TOGG,   AU_TOGG, _______, _______, _______, _______, KC_INSERT, _______, _______, KC_DELETE,
+  KC_CAPS, _______, _______,   _______, _______, AG_TOGG, _______, _______, _______,   _______, _______, _______,
+  _______, _______, KC_DELETE, _______, _______, _______, _______, _______, _______,   _______, _______, _______,
+  _______, _______, _______,   _______, _______, _______, _______, _______, _______,   _______, _______, _______
+)
+
+
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  switch (keycode) {
+        case QWERTY_MACOS:
+          if (record->event.pressed) {
+            set_single_persistent_default_layer(_QWERTY_MACOS);
+          }
+          return false;
+          break;
+        case LOWER:
+          if (record->event.pressed) {
+            layer_on(_LOWER);
+            update_tri_layer(_LOWER, _RAISE, _ADJUST);
+          } else {
+            layer_off(_LOWER);
+            update_tri_layer(_LOWER, _RAISE, _ADJUST);
+          }
+          return false;
+          break;
+        case RAISE:
+          if (record->event.pressed) {
+            layer_on(_RAISE);
+            update_tri_layer(_LOWER, _RAISE, _ADJUST);
+          } else {
+            layer_off(_RAISE);
+            update_tri_layer(_LOWER, _RAISE, _ADJUST);
+          }
+          return false;
+          break;
+        case BACKLIT:
+          if (record->event.pressed) {
+            register_code(KC_RSFT);
+            #ifdef BACKLIGHT_ENABLE
+              backlight_step();
+            #endif
+            #ifdef RGBLIGHT_ENABLE
+              rgblight_step();
+            #endif
+            #ifdef __AVR__
+            writePinLow(E6);
+            #endif
+          } else {
+            unregister_code(KC_RSFT);
+            #ifdef __AVR__
+            writePinHigh(E6);
+            #endif
+          }
+          return false;
+          break;
+      }
+    return true;
+};
+
+bool muse_mode = false;
+uint8_t last_muse_note = 0;
+uint16_t muse_counter = 0;
+uint8_t muse_offset = 70;
+uint16_t muse_tempo = 50;
+
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    if (IS_LAYER_ON(_RAISE)) {
+        if (clockwise) {
+            tap_code(KC_BRMD);
+        } else {
+            tap_code(KC_BRMU);
+        }
+    // } else if (IS_LAYER_ON(_LOWER)) { // uncomment to get a third function
+    //     if (clockwise) {
+    //         tap_code(KC_VOLD);
+    //     } else {
+    //         tap_code(KC_VOLU);
+    //     }
+    } else {
+        if (clockwise) {
+            tap_code(KC_VOLD);
+        } else {
+            tap_code(KC_VOLU);
+        }
+    }
+    return false;
+}
```

Comment:
You may want to use the encoder map feature, since it supports layers and even supports quantum and custom keycodes.

```suggestion
#if defined(ENCODER_MAP_ENABLE)
const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][2] = {
    [_QWERTY_MACOS] = { ENCODER_CCW_CW(KC_VOLU, KC_VOLD) },
    [_LOWER] = { ENCODER_CCW_CW(_______, _______) },
    [_RAISE] = { ENCODER_CCW_CW(KC_BRMU, KC_BRMD) },
    [_ADJUST] = { ENCODER_CCW_CW(_______, _______) },
};
#endif
```